File: src/main/scala/io/github/noeppi_noeppi/tools/cfupdatechecker/cache/FileCache.scala
package io.github.noeppi_noeppi.tools.cfupdatechecker.cache

import com.google.gson.{JsonArray, JsonElement, JsonObject}
import io.github.noeppi_noeppi.tools.cfupdatechecker.Util

import java.nio.file.{Files, Path, StandardOpenOption}
import scala.collection.mutable
import scala.jdk.CollectionConverters._

object FileCache {
  val VERSION: Int = 3
}

class FileCache {

  private val changelogs = mutable.Map[(Int, Int), String]()
  private val versions = mutable.Map[(Int, Int), String]()
  
  def changelog(projectId: Int, fileId: Int, changelog: => String): String = changelogs.getOrElseUpdate((projectId, fileId), changelog)
  def version(projectId: Int, fileId: Int, version: => String): String = versions.getOrElseUpdate((projectId, fileId), version)
  
  def read(path: Path): Unit = {
    
    def readFile(data: JsonObject): ((Int, Int), String) = {
      val projectId = data.get("project").getAsInt
      val fileId = data.get("file").getAsInt
      val changelog = data.get("value").getAsString
      ((projectId, fileId), changelog)
    }

    def readMap(data: JsonElement): Map[(Int, Int), String] = {
      data.getAsJsonArray.asScala.toList
        .map((e: JsonElement) => e.getAsJsonObject)
        .map(readFile)
        .toMap
    }

    changelogs.clear()
    versions.clear()
    try {
      if (Files.exists(path)) {
        val reader = Files.newBufferedReader(path)
        val json = Util.INTERNAL.fromJson(reader, classOf[JsonElement])
        reader.close()
        val cacheVersion = json.getAsJsonObject.get("version").getAsInt
        if (cacheVersion == FileCache.VERSION) {
          changelogs.addAll(readMap(json.getAsJsonObject.get("changelogs")))
          versions.addAll(readMap(json.getAsJsonObject.get("versions")))
        }
      }
    } catch {
      case e: Exception =>
        println("Failed to read file cache: " + e.getClass.getSimpleName + ": " + e.getMessage)
        changelogs.clear()
        versions.clear()
    }
  }
  
  def write(path: Path): Unit = {
    def array(elements: Seq[JsonElement]): JsonArray = {
      val array = new JsonArray()
      elements.foreach(array.add)
      array
    }
    
    def writeFile(entry: ((Int, Int), String)): JsonObject = {
      val data = new JsonObject
      data.addProperty("project", entry._1._1)
      data.addProperty("file", entry._1._2)
      data.addProperty("value", entry._2)
      data
    }

    def writeMap(map: Map[(Int, Int), String]): JsonElement = {
      array(map.toSeq.map(writeFile))
    }
    
    try {
      if (!Files.exists(path.getParent)) Files.createDirectories(path.getParent)
      val writer = Files.newBufferedWriter(path, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)
      val json = new JsonObject
      json.addProperty("version", FileCache.VERSION)
      json.add("changelogs", writeMap(changelogs.toMap))
      json.add("versions", writeMap(versions.toMap))
      writer.write(Util.INTERNAL.toJson(json) + "\n")
      writer.close()
    } catch {
      case e: Exception =>
        println("Failed to write file cache: " + e.getClass.getSimpleName + ": " + e.getMessage)
        new JsonArray
    }
  }
}

File: src/main/java/org/moddingx/cfupdatechecker/cache/FileCache.java
package org.moddingx.cfupdatechecker.cache;

import com.google.common.collect.Streams;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonSyntaxException;
import org.moddingx.cfupdatechecker.Util;

import java.io.IOException;
import java.io.Reader;
import java.io.Writer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class FileCache {

    public static final int VERSION = 4;
    private final Map<FileKey, String> changelogs = new HashMap<>();
    private final Map<FileKey, String> versions = new HashMap<>();

    public String changelog(FileKey fileKey, Supplier<String> changelog) {
        String s = this.changelogs.get(fileKey);
        if (s == null) {
            s = changelog.get();
        }

        return s;
    }

    public String version(FileKey fileKey, Supplier<String> version) {
        String s = this.versions.get(fileKey);
        if (s == null) {
            s = version.get();
        }

        return s;
    }

    public void read(Path path) {
        this.changelogs.clear();
        this.versions.clear();
        if (Files.exists(path)) {
            try (Reader reader = Files.newBufferedReader(path)) {
                JsonObject json = Util.INTERNAL.fromJson(reader, JsonObject.class);
                int cacheVersion = json.get("version").getAsInt();
                if (cacheVersion == FileCache.VERSION) {
                    this.changelogs.putAll(this.readMap(json.get("changelogs")));
                    this.versions.putAll(this.readMap(json.get("versions")));
                }
            } catch (IOException | JsonSyntaxException e) {
                System.out.println("Failed to read file cache: " + e.getClass().getSimpleName() + ": " + e.getMessage());
                this.changelogs.clear();
                this.versions.clear();
            }
        }
    }

    private Map.Entry<FileKey, String> readFile(JsonObject data) {
        int projectId = data.get("project").getAsInt();
        int fileId = data.get("file").getAsInt();
        String changelog = data.get("value").getAsString();
        return Map.entry(new FileKey(projectId, fileId), changelog);
    }

    private Map<FileKey, String> readMap(JsonElement data) {
        return Streams.stream(data.getAsJsonArray())
                .map(JsonElement::getAsJsonObject)
                .map(this::readFile)
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    public void write(Path path) {
        try {
            if (Files.notExists(path.getParent())) Files.createDirectories(path.getParent());
            try (Writer writer = Files.newBufferedWriter(path, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
                JsonObject json = new JsonObject();
                json.addProperty("version", FileCache.VERSION);
                json.add("changelogs", this.writeMap(this.changelogs));
                json.add("versions", this.writeMap(this.versions));
                writer.write(Util.INTERNAL.toJson(json) + "\n");
            }
        } catch (IOException e) {
            System.out.println("Failed to write file cache: " + e.getClass().getSimpleName() + ": " + e.getMessage());
        }
    }

    private JsonArray array(Set<JsonElement> elements) {
        JsonArray array = new JsonArray();
        elements.forEach(array::add);
        return array;
    }

    private JsonObject writeFile(Map.Entry<FileKey, String> entry) {
        JsonObject data = new JsonObject();
        data.addProperty("project", entry.getKey().projectId);
        data.addProperty("file", entry.getKey().fileId);
        data.addProperty("value", entry.getValue());
        return data;
    }

    private JsonElement writeMap(Map<FileKey, String> map) {
        JsonArray array = new JsonArray();
        map.entrySet().forEach(entry -> {
            array.add(this.writeFile(entry));
        });

        return array;
    }

    public record FileKey(int projectId, int fileId) {

    }
}
